### PR TITLE
Implements the QA system

### PR DESCRIPTION
**Pull request by dalonsoa**
It implements pre-commit hooks with the main tooling we normally used, except that we are using ruff instead of flake8 and isort. Configuration for this has been included in the `pyproject.toml` file. The workflow file has been updated to run the QA as well as all test in several OS and python versions. 

A LOT of files appear as modified since running the linters/formatters have modified them, but the functionality remains unchanged. The main files to review are:
- pyproject.toml
- ci.yml
- pre-commit-config.yaml

As the code was not following any specific style, I had to:
- disable linting the docstrings. See #33 
- disable testing in the docs and tests folders
- disable checking the line length - this needs sorting asap. See #34

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,18 +1,31 @@
...
-      uses: actions/setup-python@v2
+    - uses: actions/checkout@v4
+    - uses: actions/setup-python@v4
       with:
-        python-version: 3.10.9
+        python-version: ${{ matrix.python-version }}
+  
     - name: Install and build
-      run: pip install -r requirements-dev.txt
+      run: python -m pip install -r requirements-dev.txt
```
Comment (tsmbland) [inline .github/workflows/ci.yml:28]: Just curious what's the difference between pip install and python -m pip install?

**Comment by tsmbland** [comment]
Looks good!

**Review by tsmbland — APPROVED** [review]
Looks good

File: .github/workflows/ci.yml (hunk 1 above)
Comment (dalonsoa) [reply .github/workflows/ci.yml:28]: Have a look at https://stackoverflow.com/questions/25749621/whats-the-difference-between-pip-install-and-python-m-pip-install

We might not need it in this case, but it is becoming more common practice to ensure consistency.

**Review by AdrianDAlessandro — APPROVED** [review]
Looks good! Sensible to ignore the line lengths and docstrings for now. How did using Ruff for automatically fixing some of the problems go?

**Pull request merged**